### ISSUE TITLE
Cannot move source.tmp when installing package from git

### ISSUE THREAD

**Issue by Cutano** [xmake-io/xmake#2902]
### Xmake Version

v2.7.1

### Operating System Version and Architecture

Windows 11 22H2

### Describe Bug

![image](https://user-images.githubusercontent.com/21150043/194215946-88702b78-b296-4f58-96b6-cbbc67f6892d.png)

error: cannot move source.tmp to source Permission denied

### Expected Behavior

Deps installed successfully.

**Comment by waruqi** [comment]
you can add `-vD` and let me see full logs.

```
xmake f -c -vD
```

**Comment by Cutano** [comment]
Thanks for the reply!  
Here's the full logs:
![image](https://user-images.githubusercontent.com/21150043/194486122-67ed0f20-0077-4231-a8ee-b8bff1e4dbfe.png)

**Comment by Cutano** [comment]
Here's another log when I use particular version of ImGui:
![image](https://user-images.githubusercontent.com/21150043/194498892-f66b6c29-78f7-4fa0-89d3-d99c9c515ba1.png)

![image](https://user-images.githubusercontent.com/21150043/194498804-e296748b-7768-487c-b712-cb16927e4839.png)

**Comment by waruqi** [comment]
I don't know, maybe some other process is accessing this directory, you can check it. `%localappdata%/.xmake/cache/... /source.tmp`

**Comment by Cutano** [comment]
It seems that the problem is associated with Git. I cleared my Git config then reinstalled it, and it's fine.

**Comment by xq114** [comment]
Please reopen this issue. I just ran into the same issue; I'm wondering why xmake clone a repository to `source.tmp` instead of copying it directly into `source`?

https://github.com/xmake-io/xmake/blob/4139d07bac797f7775d6a3957a1aa37c67d730ff/xmake/modules/private/action/require/impl/actions/download.lua#L78-L102

I think it does nothing but making more errors, since git could occupy this folder, and the performance would suffer from the useless copying.

**Comment by SirLynix** [comment]
> and the performance would suffer from the useless copying
Moving a folder on the same disk is a very fast operation, it doesn't involve copying.

My theory on this error is that git is still using the folder after cloning it on some configs, maybe to index it, so it locks the .tmp folder preventing it to be moved.

**Comment by waruqi** [comment]
os.mv is very fast, and normally, source.tmp should only be occupied by one git process, and it will stop occupying this directory as soon as git clone finishes. Unless the git process or its child processes have not fully exited after git clone has finished and continue to occupy it.

The use of source.tmp is to ensure the integrity of the source directory, so that even if git is killed during the clone process, there will not be an incomplete source directory.

**Comment by SirLynix** [comment]
maybe xmake could try to move it multiple times if it fails, waiting a bit between tries until it reaches a max_tries value.

**Comment by waruqi** [comment]
But I need to know the root cause, why the git clone process exits and source.tmp continues to be occupied.

**Comment by SirLynix** [comment]
> I have tried rebooting and build directly after that, there should be no process occupied, but failed with the same issue.

if that operation failed xmake will clone it again with git, causing the same issue.

If someone successfully reproduces this issue, it would be nice to use something like https://lockhunter.com to quickly check what processes lock the .tmp folder

**Comment by waruqi** [comment]
> > os.mv is very fast, and normally, source.tmp should only be occupied by one git process, and it will stop occupying this directory as soon as git clone finishes. Unless the git process or its child processes have not fully exited after git clone has finished and continue to occupy it.
> 
> I have tried rebooting and build directly after that, there should be no process occupied, but failed with the same issue.

That's weird, it's like a MoveFileExW problem, but I don't know why it's failing if no any process occupied.

**Comment by waruqi** [comment]
@Cutano You can check that no other process is occupying it, or run `xmake l os.mv xxxx/source.tmp xxxx/source` to quickly verify that os.mv is working

**Comment by Cutano** [comment]
> @Cutano You can check that no other process is occupying it, or run `xmake l os.mv xxxx/source.tmp xxxx/source` to quickly verify that os.mv is working

I can't reproduce after I reinstalled git. Btw, is there anything to do with the Detached HEAD state of the cloned git repo?
![image](https://user-images.githubusercontent.com/21150043/194704782-f75521e7-3f45-4ad2-9e6f-a8a61d9a65a9.png)

**Comment by waruqi** [comment]
> > @Cutano You can check that no other process is occupying it, or run `xmake l os.mv xxxx/source.tmp xxxx/source` to quickly verify that os.mv is working
> 
> I can't reproduce after I reinstalled git. Btw, is there anything to do with the Detached HEAD state of the cloned git repo? ![image](https://user-images.githubusercontent.com/21150043/194704782-f75521e7-3f45-4ad2-9e6f-a8a61d9a65a9.png)

It shouldn't have anything to do with this.

**Comment by xq114** [comment]
![image](https://user-images.githubusercontent.com/42310255/194705947-d7c86e91-1b8f-4d87-9c5e-7d2c21b4047b.png)

**Comment by xq114** [comment]
只要xmake不退出，git也不退出，我在中间加了sleep 20s，git就一直占用了20s

**Comment by waruqi** [comment]
Where did you add os.sleep? As soon as the os.exec("git clone") call returns, the git process should have exited

**Comment by SirLynix** [comment]
[Process Monitor](https://learn.microsoft.com/en-us/sysinternals/downloads/procmon) could help understand what's going on here, it's very powerful.

**Comment by xq114** [comment]
![image](https://user-images.githubusercontent.com/42310255/194706185-8ec12b7e-9c48-42f0-b412-402124de817e.png)
here

**Comment by waruqi** [comment]
> ![image](https://user-images.githubusercontent.com/42310255/194706185-8ec12b7e-9c48-42f0-b412-402124de817e.png) here

The git.clone/git.checkout calls have returned and the git process should have exited. I don't know why the git process still exists.

**Comment by xq114** [comment]
And there is another strange thing. Spawning only one git process will not cause the issue.

This code will cause the issue;
![image](https://user-images.githubusercontent.com/42310255/194706992-e934b763-396b-4dbd-a4ef-06d4049f81dc.png)

While this won't.
![image](https://user-images.githubusercontent.com/42310255/194707016-079d06e9-ee0e-4bd1-86f5-8bc4abba496e.png)

**Comment by xq114** [comment]
if the issue cannot be addressed, can xmake provide a way to get around? i.e. force xmake to treat source.tmp as complete

**Comment by waruqi** [comment]
no way. And if another git process does keep occupying source.tmp, then even cloning directly to the source directory may cause other problems.

**Comment by waruqi** [comment]
I tried it, it works for me.

```lua
import("devel.git")

function main()
    os.tryrm("testgit")
    git.clone("https://github.com/nothings/stb.git", {outputdir = "testgit"})
    git.checkout("4af130e86341928e3003ba5657f3e9faec50c1dc", {repodir = "testgit"})
    os.tryrm("teststb")
    os.mv("testgit", "teststb")
end
```

**Comment by waruqi** [comment]
@xq114 Can you try using xmake's built-in git instead of mingw/git? Does this problem still exist?

**Comment by waruqi** [comment]
> ![image](https://user-images.githubusercontent.com/42310255/194705947-d7c86e91-1b8f-4d87-9c5e-7d2c21b4047b.png)

where is your git.exe? my git is `mingw64\bin\git.exe`, it will call some git processes in `mingw64\libexec\git-core`

but your git process is `mingw64\libexec\git-core\git.exe`

It is possible that the git process called by xmake has already exited, but an additional git.exe child process has been created inside git, which is associated with the parent process and has not been exited.

**Comment by xq114** [comment]
git 2.34.1 is also working

**Comment by xq114** [comment]
git 2.32.0 did not work. Should xmake prompt a warning if their git is outdated?

**Comment by waruqi** [comment]
> git 2.32.0 did not work. Should xmake prompt a warning if their git is outdated?

But xmake doesn't know if it's outdated. Theoretically, xmake can use any version of git until it runs into this problem

**Comment by xq114** [comment]
> > git 2.32.0 did not work. Should xmake prompt a warning if their git is outdated?
> 
> But xmake doesn't know if it's outdated. Theoretically, xmake can use any version of git until it meet this problem
> 
> and my git is 2.20.0, it works.

Did you test it on windows? Since now only windows users reported the issue

**Comment by waruqi** [comment]
> > > git 2.32.0 did not work. Should xmake prompt a warning if their git is outdated?
> > 
> > 
> > But xmake doesn't know if it's outdated. Theoretically, xmake can use any version of git until it meet this problem
> > and my git is 2.20.0, it works.
> 
> Did you test it on windows? Since now only windows users reported the issue

yeah, on my win7

**Comment by waruqi** [comment]
@Cutano what is your git version?

**Comment by waruqi** [comment]
> git 2.32.0 did not work. Should xmake prompt a warning if their git is outdated?

I tried 2.32.2. it works for me. https://github.com/git-for-windows/git/releases/download/v2.32.0.windows.1/MinGit-2.32.0-64-bit.zip

```lua
import("devel.git")

function main()
    os.tryrm("testgit")
    git.clone("https://github.com/nothings/stb.git", {outputdir = "testgit"})
    git.checkout("4af130e86341928e3003ba5657f3e9faec50c1dc", {repodir = "testgit"})
    os.tryrm("teststb")
    os.mv("testgit", "teststb")
end
```

```console
xmake l test.lua
```

**Comment by Cutano** [comment]
> @Cutano what is your previous git version?

Probably 2.30, I deleted the package but I remember git was installed around May.

**Comment by Longxr** [comment]
![source tmp_20221017145323](https://user-images.githubusercontent.com/15936923/196115168-319a9aa0-1d0d-4286-9071-cac7cafe71be.png)

Xmake Version
v2.5.6
Operating System Version and Architecture
Windows 11 21H2
Git Version
git version 2.34.0.windows.1 did not work
git version 2.38.0.windows.1 also did not work

**Comment by kantraksel** [comment]
I had similar problem (could not install sentry-native, because source.tmp was busy), but i remembered git had introduced fsmonitor. So I disabled fsmonitor (`git config --global core.fsmonitor true`) and killed all remaining git processes (which were instances of fsmonitor). After that xmake did all remaining work.

Win11 with latest updates, git version 2.38.1.windows.1

**Comment by waruqi** [comment]
> I had similar problem (could not install sentry-native, because source.tmp was busy), but i remembered git had introduced fsmonitor. So I disabled fsmonitor (`git config --global core.fsmonitor false`) and killed all remaining git processes (which were instances of fsmonitor). After that xmake did all remaining work.
> 
> Win11 with latest updates, git version 2.38.1.windows.1

@Longxr @xq114 @Cutano Maybe this is the cause of the problem, you can try.

**Comment by waruqi** [comment]
I have improved it, can you try it?

xmake update -s github:xmake-io/xmake#fswatcher

**Comment by xq114** [comment]
> ![source tmp_20221017145323](https://user-images.githubusercontent.com/15936923/196115168-319a9aa0-1d0d-4286-9071-cac7cafe71be.png)
> 
> Xmake Version v2.5.6 Operating System Version and Architecture Windows 11 21H2 Git Version git version 2.34.0.windows.1 did not work git version 2.38.0.windows.1 also did not work
> 
> edit C:\Program Files\xmake\modules\private\action\require\impl\actions\download.lua os.mv => os.cp can work ![image](https://user-images.githubusercontent.com/15936923/196134634-1141e76c-1ec9-4b0e-a3ab-65069d47f9b6.png)

Then how about this one? This issue happens with curl, not git

**Comment by waruqi** [comment]
> > ![source tmp_20221017145323](https://user-images.githubusercontent.com/15936923/196115168-319a9aa0-1d0d-4286-9071-cac7cafe71be.png)
> > Xmake Version v2.5.6 Operating System Version and Architecture Windows 11 21H2 Git Version git version 2.34.0.windows.1 did not work git version 2.38.0.windows.1 also did not work
> > edit C:\Program Files\xmake\modules\private\action\require\impl\actions\download.lua os.mv => os.cp can work ![image](https://user-images.githubusercontent.com/15936923/196134634-1141e76c-1ec9-4b0e-a3ab-65069d47f9b6.png)
> 
> Then how about this one? This issue happens with curl, not git

I don't know, but it should have nothing to do with curl, only unpacking tar.gz will access source.tmp. But 7z should not take up source.tmp when it finishes unpacking the file

**Comment by waruqi** [comment]
> I have improved it, can you try it?
> 
> ```
> xmake update -s github:xmake-io/xmake#fswatcher
> ```

Have you all tested this patch and if it works I plan to merge it.

**Comment by waruqi** [comment]
I've merged the patch for now, so you can watch it for a while

**Comment by Cutano** [comment]
> > I have improved it, can you try it?
> > ```
> > xmake update -s github:xmake-io/xmake#fswatcher
> > ```
> 
> Have you all tested this patch and if it works I plan to merge it.

Works for me, seems to be the root cause, thanks!

**Comment by waruqi** [comment]
I have merged into dev.

I closed it first and if you have this questions later, you can open it again.

**Comment by waruqi** [comment]
> > I have tried rebooting and build directly after that, there should be no process occupied, but failed with the same issue.
> 
> if that operation failed xmake will clone it again with git, causing the same issue.
> 
> If someone successfully reproduces this issue, it would be nice to use a software like https://lockhunter.com to quickly check what processes lock the .tmp folder

another tool, handle.exe

https://learn.microsoft.com/zh-cn/sysinternals/downloads/handle

we can run `os.exec("handle.exe")` to dump all when os.rm is failed.

**Comment by waruqi** [comment]
> Then how about this one? This issue happens with curl, not git

https://github.com/tboox/tbox/blob/5832d9d3eb26f4472de8f1f4de33a3367642c9eb/src/tbox/platform/windows/process.c#L452

```c
        bInheritHandle = TRUE;
        if (!tb_kernel32()->CreateProcessW(tb_null, command, &sap, &sat, bInheritHandle, flags, (LPVOID)environment, attr && attr->curdir? curdir : tb_null, &process->si, &process->pi))
```

I probably figured out the root cause. xmake creates child processes with inheritance handles enabled by default.

If source.tmp is already open and occupied (os.cd()) when multiple subprocesses are executed in parallel, they will all inherit its handle.

Even if the current git/unzip subprocess has exited and no longer accesses it, it will continue to be occupied by other subprocesses that may not have direct access to it.

I had a similar problem when I executed the cl.exe sub-process in parallel and I was unable to delete the files accessed by other cl.exe processes.

But if I disable the inheritance handle, it works.

https://github.com/xmake-io/xmake/blob/a6e149fc533f1ffb7cd7b4bd663caa7bb6b0be35/xmake/modules/core/tools/cl.lua#L512

**Comment by waruqi** [comment]
Although I know the cause of the problem, I don't know how to fix it and I can't disable handle inheritance. I can't disable handle inheritance because it is needed for the process to redirect input and output.

Why can't CreateProcess be configured directly to inherit only the STARTUPINFO.hStdOutputhandle of the current incoming parameter?

Tell CreateProcess to inherit by SetHandleInformation(si.hStdOutput, HANDLE_FLAG_INHERIT, TRUE) and it will globally affect other child processes.

- https://gitlab.haskell.org/ghc/ghc/-/issues/2650
- https://forums.codeguru.com/showthread.php?432338-Child-process-inheriting-all-handles

**Comment by waruqi** [comment]
I have improve to spawn process on windows. https://github.com/tboox/tbox/commit/5b62232dce4302e512a659ecb2403c312e0b9cb7

**Comment by waruqi** [comment]
I found that if the directory contains files with long paths, even if the original directory itself has a short path, MoveFileEx will also fail and return a permission error (lasterror: 5).
